### PR TITLE
Action bar idea

### PR DESCRIPTION
**Pull request by bentleyvk**
image

It is obviously not finished. Need to discuss whether action bar should be a placeholder or have full styling and centering. What we should do with content, should we center it, limit max-width?

### CODE REVIEW

**Comment by bentleyvk** [comment]
@FlyersPh9 @gretanausedaite would like to get your opinion about this.

**Review by gretanausedaite — COMMENTED** [review]
It looks good from code perspective to me.
We are giving users full control on background and content.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            }}
+          >
+            <LabeledInput label='Name' />
+            <LabeledInput label='Description' />
+            <LabeledInput label='Country' />
+            <LabeledInput label='City' />
+          </form>
+        </PageLayout.Content>
+        <PageLayout.ActionBar>
+          <Surface
```
Comment (FlyersPh9) [inline demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:74]: Can we put the visual styling on `.iui-layouts-page-content-action-bar` & remove `<Surface>`?  I think [this new `<Flex>` component](https://github.com/iTwin/iTwinUI/pull/1011) will help us achieve the centering & gap.

Comment (FlyersPh9) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:74]: To expand upon this, we should probably copy the horizontal padding rules that are applied to the body content area.  If the side nav is removed, the body content includes padding to accommodate for the notch & rounded edges of the screen.
Screenshot 2023-01-20 at 12 03 07 PM
Screenshot 2023-01-20 at 12 05 57 PM

**Comment by FlyersPh9** [comment]
This misalignment is not great, but at the same time we should use our CSS spacing vars & not use some weird number.  Maybe unavoidable, but maybe not, will need to investigate.
Screenshot 2023-01-20 at 12 09 38 PM

Also, maybe the top border should be 2px to match the side nav & header.

**Comment by bentleyvk** [comment]
There is one concern that `ActionBar` is a sibling to `Content` so if users have some kind of page template component that looks something like this:
```tsx
// Template.tsx
const PageTemplate = ({ children }) => {
  return (
    <PageLayout>
        <PageLayout.Header>
          <DemoHeader />
        </PageLayout.Header>
        <PageLayout.Content padded>
           {children}
        </PageLayout.Content>
    </PageLayout>
  )
}
```
They will have problems using our `ActionBar` since they won't have any control other it. So the best way might be having action bar inside content.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 1 above)
Comment (FlyersPh9) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: I added the notch padding to the action bar in [ca33911](https://github.com/iTwin/iTwinUI-layouts/pull/70/commits/ca339113fb9004b2bf76b47f41cbb085e8a27d5f).

**Comment by LostABike** [comment]
@gretanausedaite it looks like the current layout with `ActionBar` is already of this structure 
```

<Button bar>

  <Left side></Left side>

  <Middle part></Middle part>

  <Right side></Right side>

</Button bar>

```
Do we need or want to add stylings to left, middle and right part?

I took @bentleyvk suggestion and tried to make `ActionBar` child of `Content` which requires some css changes to keep the position still at bottom, but it seems wrong (tested with one case adding `left` for small-monitor `      @include small-monitor {
        padding-inline: unquote('max(calc(var(--iui-size-l) * 2), env(safe-area-inset-right))') unquote('max(calc(var(--iui-size-l) * 2), env(safe-area-inset-left))');
        left: calc(var(--iui-size-l) *3);
      }`)
      
Maybe there is a better way to do this.
There is also that issue with `ActionBar` height misaligned with the side nav buttons height, I'm not sure how we can fix that (without passing heights of other comps around but then that would make them dependent on each other). Or maybe we could find and use the same height calculation for side nav bottom part.

**Comment by LostABike** [comment]
Reverted previous change and updated the component instead. Also made the height for the bar to be `calc(calc(var(--iui-size-s) * 5) + var(--iui-size-l));` because that is the size of the two buttons bottom of sidenav. Not ideal and still slightly misaligned lol.

**Comment by gretanausedaite** [comment]
I think it would be the best if we supported these scenarios:

Just a slot at the bottom of the page without any background, paddings, borders, anything.
```js
<PageLayout.BottomBar>
  {content}
</PageLayout.BottomBar>
```

Adding ActionBar (as a separate component) which adds background, border and padding to the area, fully custom user content
```js
<PageLayout.BottomBar>
  <ActionBar>{content}</ActionBar>
</PageLayout.BottomBar> 
```

Fully predesigned layout - slots for left, center and right areas
```js
<PageLayout.BottomBar>
  <ActionBar>
    <ActionBar.Left>{content}</ActionBar.Left>
    <ActionBar.Center>{content}</ActionBar.Center>
    <ActionBar.Right>{content}</ActionBar.Right>
  </ActionBar>
</PageLayout.BottomBar>
```

To achieve this you'll need to create new component BottomBar and add it as subcomponent of PageLayout. This component should only have grid-area and height defined.

Adjust ActionBar to include background, border, padding and create subcomponents for Left, Right and Center areas. 

Let me know if this make sense.

**Comment by LostABike** [comment]
@gretanausedaite as mentioned, I have reverted all the old changes and implemented all the components that you suggested. Which includes BottomBar, ActionBar and its subcomponents: Left, Center, Right. All the stylings were done in `page.scss` and demo in `PageLayoutStickyBarDemo.tsx`.

File: packages/layouts-react/src/components/PageLayout/Left.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
```
Comment (gretanausedaite) [inline packages/layouts-react/src/components/PageLayout/Left.tsx:5]: You can combine all ActionBar subcomponents into ActionBar file. It would take less space (eg. what we are doing with Flex, FileUploadCard)

Comment (LostABike) [reply packages/layouts-react/src/components/PageLayout/Left.tsx]: I have combined all ActionBar subcomponents and deleted the old ones.

File: packages/layouts-react/src/components/PageLayout/PageLayout.tsx [hunk 3]
```diff
@@ -33,6 +34,7 @@ PageLayout.SideNavigation = SideNavigation;
 PageLayout.Content = Content;
 PageLayout.ToolsArea = ToolsArea;
 PageLayout.TitleArea = TitleArea;
+PageLayout.BottomBar = BottomBar;
 PageLayout.ActionBar = ActionBar;
```
[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: I have moved `ActionBar` into its own folder under `components`. I think the css remains the same? Since it is dependent on the demo layout which has `ActionBar` under `BottomBar` which is subcomponent of `PageLayout`.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 4]
```diff
@@ -70,20 +74,26 @@ export const PageLayoutStickyBarDemo = () => {
...
+            <PageLayout.ActionBar.Center>
+              <div>
+                <Button styleType='cta' size='large'>
+                  Create
+                </Button>
+                <Button size='large'>Cancel</Button>
+              </div>
+            </PageLayout.ActionBar.Center>
+            <PageLayout.ActionBar.Right>
+              <div />
```
[the remark this answers is not in the document]
Comment (LostABike) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: I have removed this.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 5]
```diff
@@ -35,6 +35,10 @@ const Demo = () => {
           <DemoContent />
         </Surface>
       </PageLayout.Content>
+
+      
+      <PageLayout.BottomBar>
```
[the remark this answers is not in the document]
Comment (gretanausedaite) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:40]: Yup

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (gretanausedaite) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: Remove whole <ActionBar.Right /> component.
When you do that, you might notice that layout shifts, it should not happen. You'll need to fix css to not require any of the areas (left, center or right) so users can choose which to include in their apps.

File: packages/layouts-react/src/components/PageLayout/PageLayout.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (gretanausedaite) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: I'd like to remove action bar from under bottom bar (at least).
Also BottomBar doesn't need background - leave it completely empty, just height and area in the grid.

File: packages/layouts-css/src/page.scss [hunk 6]
```diff
@@ -17,8 +18,7 @@
     grid-area: sidenav;
 
     //#region If side navigation is present, remove the env(safe-area-inset-left) from the left side of content area
-    ~ .iui-layouts-page-content-padded {
-      padding-left: var(--iui-size-2xl);
+    ~ .iui-layouts-page-content-padded, {
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:21]: remove `,`

File: packages/layouts-css/src/page.scss [hunk 7]
```diff
@@ -46,22 +46,97 @@
...
+        padding: var(--iui-size-xs)
+          unquote('max(var(--iui-size-l), env(safe-area-inset-right))')
+          var(--iui-size-xs)
+          unquote('max(var(--iui-size-l), env(safe-area-inset-left))');
       }
       //#endregion Add env(safe-area-inset-*) for iPhone notch on the left & right sides of content area
     }
 
+
+    &-tools {
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:78]: Why did you add tools here?

File: packages/layouts-css/src/page.scss [hunk 8]
```diff
@@ -46,22 +46,97 @@
...
+    @include tablet {
+      padding-inline: unquote('max(var(--iui-size-xl), env(safe-area-inset-right))') unquote('max(var(--iui-size-xl), env(safe-area-inset-left))');
+    }
+
+    @include mobile {
+      padding: unquote('max(var(--iui-size-l), env(safe-area-inset-right))') unquote('max(var(--iui-size-l), env(safe-area-inset-left))');
+    }
+    //#endregion Add env(safe-area-inset-*) for iPhone notch on the left & right sides of content area
+
+    &-zones {
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:116]: What "-zones" are for?

Comment (LostABike) [reply packages/layouts-css/src/page.scss:116]: Oops that is leftover styling of previous layout. Removed.

File: packages/layouts-css/src/page.scss (hunk 7 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss:78]: It was there before. I think it is for the styling of `ToolsArea.tsx` component that is used in layout demo 2.

File: packages/layouts-css/src/page.scss (hunk 6 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss:21]: Thanks. I removed this.

File: packages/layouts-react/src/components/ActionBar/ActionBar.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+// ----------------------------------------------------------------------------
+// ActionBar.Left component
+export type LeftProps = {
+  /**
+   * Children of the element
+   */
+  children: React.ReactNode;
+} & StylingProps;
+
+const Left = (props: LeftProps) => {
```
[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/ActionBar/ActionBar.tsx:18]: I added jsdocs, but not too confident on the format. I have read how these are written in other components in our repo but they seem varied.

File: packages/layouts-react/src/components/ActionBar/ActionBar.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    <div
+      className={cx('iui-layouts-page-bottom-bar-action-bar', className)}
+      style={style}
+    >
+      {children}
+    </div>
+  );
+};
+
+ActionBar.displayName = 'PageLayout.ActionBar';
```
[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/ActionBar/ActionBar.tsx:97]: Thanks! I removed this.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (LostABike) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: You're right, I thought `width: 33%` would be sufficient but it is not. I have removed that and set a grid display instead so each component go into their appropriate slot on the grid.

File: packages/layouts-css/src/page.scss [hunk 11]
```diff
@@ -46,22 +46,97 @@
...
+
+      &-right {
+        justify-content: flex-end;
+      }
+    }
+
+
+    //#region Add env(safe-area-inset-*) for iPhone notch on the left & right sides of content area
+
+    @include small-monitor {
```
[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-css/src/page.scss:103]: You are right. They were there prior, and was part of `-action-bar`, I must have missed them when I moved `-action-bar` out. I moved them back.

File: packages/layouts-react/src/components/PageLayout/PageLayout.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: I'm still working on adjusting the css for this because moving it out is messing up the whole interface that previously relied on being under `BottomBar` and `PageLayout`. But just to make absolutely sure I understand correctly, do you mean for ActionBar to be placed like this?
image
and still rely on page layout grid to position at the bottom.

Or like this:
image
Where it is completely outside of PageLayout and BottomBar, and will have independent positioning (to be fixed at the bottom)? and its own styling.

[the remark this answers is not in the document]
Comment (gretanausedaite) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: Looking back to my initial examples: https://github.com/iTwin/iTwinUI-layouts/pull/70#issuecomment-1517594060

```ts
<PageLayout>
  <PageLayout.Content>
    {content}
  </PageLayout.Content>
  <PageLayout.BottomBar>
    {bottom bar content, might be ActionBar}
  </PageLayout.BottomBar>
</PageLayout>
```

[the remark this answers is not in the document]
Comment (gretanausedaite) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: CSS should be something like:
```css
.iui-layouts-page {
  &-bottom-bar {
    grid-area: bottom-bar;
    height: calc(var(--iui-size-s)*5 + var(--iui-size-l)); //this is the sizes of the two bottom buttons of side nav
  }

  &-action-bar {
    [action bar css code]
  }
}
```

or 
```css
.iui-layouts-page {
  &-bottom-bar {
    grid-area: bottom-bar;
    height: calc(var(--iui-size-s)*5 + var(--iui-size-l)); //this is the sizes of the two bottom buttons of side nav
  }
}

.iui-action-bar {
  [action bar css code]
}
```

[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: Okay I updated this.

File: packages/layouts-react/src/components/PageLayout/Right.tsx [hunk 12]
```diff
@@ -0,0 +1,30 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (gretanausedaite) [inline packages/layouts-react/src/components/PageLayout/Right.tsx:1]: Is this leftover file?

File: packages/layouts-react/src/components/PageLayout/Center.tsx [hunk 13]
```diff
@@ -0,0 +1,30 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (gretanausedaite) [inline packages/layouts-react/src/components/PageLayout/Center.tsx:1]: Is this leftover file?

File: packages/layouts-react/src/components/PageLayout/PageLayout.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (gretanausedaite) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: Did you push these changes?I I'm not seeing it.

[the remark this answers is not in the document]
Comment (LostABike) [reply packages/layouts-react/src/components/PageLayout/PageLayout.tsx]: My bad, it should show now.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 14]
```diff
@@ -0,0 +1,100 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (gretanausedaite) [inline demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:1]: Add test for this file to get screenshots.

File: packages/layouts-css/src/page.scss [hunk 15]
```diff
@@ -94,5 +96,71 @@
         justify-content: flex-end;
       }
     }
+
+    &-centered {
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:100]: Move back to original place (to avoid unnecessary changes)

File: packages/layouts-css/src/page.scss [hunk 16]
```diff
@@ -94,5 +96,71 @@
...
+    height: calc(var(--iui-size-s)*5 + var(--iui-size-l)); //this is the sizes of the two bottom buttons of side nav
+  }
+
+  &-action-bar {
+    background-color: var(--iui-color-background);
+    border-top: 2px solid var(--iui-color-border);
+    margin-inline: auto;
+    padding: var(--iui-size-m);
+    height: inherit;
+    display: grid;
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:123]: Add small `gap` to create separation between areas if users add a lot of things into them

File: packages/layouts-css/src/page.scss [hunk 17]
```diff
@@ -94,5 +96,71 @@
...
+    grid-template-areas: 'left center right';
+
+    //#region Add env(safe-area-inset-*) for iPhone notch on the left & right sides of content area
+
+    @include small-monitor {
+      padding-inline: unquote('max(calc(var(--iui-size-l) * 2), env(safe-area-inset-right))') unquote('max(calc(var(--iui-size-l) * 2), env(safe-area-inset-left))');
+    }
+
+    @include tablet {
+      padding-inline: unquote('max(var(--iui-size-xl), env(safe-area-inset-right))') unquote('max(var(--iui-size-xl), env(safe-area-inset-left))');
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:134]: Maybe switch template to vertical display to make it fit into page.
You might also need to adjust initial height and justify-content settings in areas.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+        </PageLayout.Content>
+        <PageLayout.BottomBar>
+          <ActionBar>
+            <ActionBar.Left>
+              <div>
+                <Checkbox label='Remember my settings' />
+              </div>
+            </ActionBar.Left>
+            <ActionBar.Center>
+              <div>
```
Comment (gretanausedaite) [inline demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:86]: Add small gap between buttons for better display.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            <LabeledInput label='Name' />
+            <LabeledInput label='Description' />
+            <LabeledInput label='Country' />
+            <LabeledInput label='City' />
+          </form>
+        </PageLayout.Content>
+        <PageLayout.BottomBar>
+          <ActionBar>
+            <ActionBar.Left>
+              <div>
```
Comment (gretanausedaite) [inline demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:81]: I think you don't need this additional `div` wrap.

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 14 above)
Comment (LostABike) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx:1]: I have added test

File: packages/layouts-css/src/page.scss (hunk 15 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss]: I have moved this back

File: packages/layouts-css/src/page.scss (hunk 16 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss:122]: Added gap with size `--iui-size-xs`

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 19 above)
Comment (LostABike) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: Right, I have removed this, and everything looks the same.

File: packages/layouts-css/src/page.scss (hunk 17 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss:134]: I don't really understand what you mean by switching template to vertical display. Could you elaborate more? Do you mean for only `tablet` case?

File: packages/layouts-css/src/page.scss [hunk 20]
```diff
@@ -95,4 +107,61 @@
...
+    }
+
+    @include mobile {
+      padding: unquote('max(var(--iui-size-l), env(safe-area-inset-right))') unquote('max(var(--iui-size-l), env(safe-area-inset-left))');
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:138]: You probably need to update padding to match padded content padding. Right now on mobile padding takes almost all available space. 
image

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 18 above)
Comment (gretanausedaite) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: We should not assume what content users are putting here and what gap they would need.
This gap setting should go in the demo page.

File: packages/layouts-css/src/page.scss (hunk 17 above)
Comment (gretanausedaite) [reply packages/layouts-css/src/page.scss:134]: You could add something like this to mobile version of action bar: 
image
image

It would make grid to display items in one column:
image

@FlyersPh9 if that is ok?

File: demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx (hunk 18 above)
Comment (LostABike) [reply demo/src/demos/PageLayout/PageLayoutStickyBarDemo.tsx]: Thanks that makes sense. I've added gap in demo page only and removed gap in `page.scss`

File: packages/layouts-css/src/page.scss (hunk 20 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss]: Right yea. I've adjusted it to size s and it looks better now.

File: packages/layouts-css/src/page.scss (hunk 17 above)
Comment (LostABike) [reply packages/layouts-css/src/page.scss:138]: Thanks for the guide. I have updated this. Only thing would be that it no longer aligns with the side nav buttons but I think ultimately the specifics of content will be up to the users to adjust accordingly.

File: packages/layouts-css/src/page.scss [hunk 21]
```diff
@@ -95,4 +107,69 @@
...
+    @include mobile {
+      height: fit-content;
+    }
+  }
+
+  &-action-bar {
+    background-color: var(--iui-color-background);
+    border-top: 2px solid var(--iui-color-border);
+    margin-inline: auto;
+    padding: var(--iui-size-m);
```
Comment (gretanausedaite) [inline packages/layouts-css/src/page.scss:124]: Padding on bigger screens seems to be smaller than in mid size screens. Maybe this could work.

```suggestion
    padding-inline: unquote('max(calc(var(--iui-size-xl) * 2), env(safe-area-inset-right))') unquote('max(calc(var(--iui-size-xl) * 2), env(safe-area-inset-left))');

```

File: packages/layouts-css/src/page.scss (hunk 20 above)
Comment (gretanausedaite) [reply packages/layouts-css/src/page.scss]: size m could also work. S also sounds good.

**Review by gretanausedaite — APPROVED** [review]
Looks good to me.
@FlyersPh9 could you check the visuals?

File: packages/layouts-css/src/page.scss (hunk 17 above)
Comment (FlyersPh9) [reply packages/layouts-css/src/page.scss:138]: I'm fine with this for now.  

We don't have a great solution for mobile at the moment.  A while back we explored replacing the side navigation with a bottom bar in #34, if we ever did implement that the double bottom bar will be kind of strange.  But for where we are now, this suggestion makes sense to me.

**Review by FlyersPh9 — APPROVED** [review]
I checked this out in Brave (Chrome) & Safari on macOS and iOS Simulator.  I tried it with and without the side nav present.  Looks great, ship it!
